### DISCUSSION TITLE
Render Delegation하는 React 컴포넌트 만들기 | kciter.so

### DISCUSSION THREAD

**Discussion by giscus** [kciter/kciter.github.io#28] [Announcements]
# Render Delegation하는 React 컴포넌트 만들기 | kciter.so

이전에  Polymorphic한 React 컴포넌트 만들기 라는 글을 작성했다. 해당 글에서는  as  속성을 이용하여 컴포넌트를 변형시키는 것을 소개하는데 이를 이용하면 하나의 컴포넌트로 여러 요소(Element)나 그에 따른 속성(Attribut…

https://kciter.so/posts/render-delegation-react-component

**Comment by jiwoo-choi** [comment]
지난 Polymorphic 컴포넌트 글을 보고 회사 시스템에도 적용해봤는데요 :) 
직접 사용해보니 Polymorphic 컴포넌트로만은 리액트 컴포넌트 운용에 한계가 있더라구요. 
그래서 Rendering Delegation도 일부 추가해서 사용해보았고, 두가지 방법을 사용하면서 느낀점을 간단히 공유해봅니다!

### polymorphic

polymorhpic은 단순하고 투명하다라는 점이 좋았어요. 
런타임에서 해결시키는 기능들이 적고, 주로 props(i.e.As Props)로 소통 해요. 그래서 코드가 단순해지고 투명해져요. 작업자 누구나 as props가 어떤식으로 흘러가는지 쉽게 확인할 수 있는게 큰 장점이었어요.

만약 다른 프로젝트에서 도입을 해본다면 아래 내용들을 고려해볼 것 같아요.
- 컴포넌트의 껍데기에 다형적인 기능을 추가하는 것 외의 용도로 사용하려고 하면 코드관리가 어렵고, 특유의 순수성도 많이 헤치는 것 같아요. 그래서 atomic(예: header, button) 레벨의 요소에 적용해두는게 가장 적합해요.
- typescript를 사용한다면 컴포넌트 정의부에서 복잡한 타입을 정의해놔야해요. 작업자마다 타입에 대한 이해수준이 다를 수 있어서 거부감을 느낄 수 있어요. 저희팀은 MUI의 다형타입 유틸타입을 참고하여 다형타입을 정의해주는 유틸타입을 만들었어요.

### Rendering Delegation
rendering delegation은 런타임 기능들이 많아서 컴포넌트의 조합을 유연하고 우아하게 처리할 수 있는게 좋았어요.


만약 다른 프로젝트에서 도입을 해본다면 아래 내용들을 고려해볼 것 같아요.
- 컴포넌트의 기능이 완전히 정의된 상태에서는 이 방법이 유용해요. 하지만, 만약 기능 정의가 안된 컴포넌트에 무작정 asChild를 열어두면 혼란함을 야기할 것 같아요. 코드가 어떻게 적용되는지 불투명하다보니 '기준 기능 명세'가 없으면 불안해요.
- slottable한 시스템이 꼭 필요할지? : 생각보다 컴포넌트의 기능을 정의해두고 만드는 케이스가 많지 않아요 (팀마다, 디자인시스템 정책마다 다를 수 있어요). 그렇다면 굳이 공통화된 slottable한 시스템을 구축하기보단, 위 글에서 언급해주신 방법론들을 적절히 조합해서 사용하는게 더 유리할 수 있어요.

**Reply by kciter** [reply]
상세하게 정리해주셔서 감사합니다. 굉장히 늦게 후속글을 작성했는데 기억해주시니 기쁘네요 😄
애매한 지점을 정확하게 짚어주셔서 저를 비롯해서 도입을 고민하시는 분들에게 크게 도움이 될 것 같습니다.
특히 `polymorhpic은 단순하고 투명하다`, `rendering delegation은 컴포넌트의 기능이 완전히 정의된 상태에 유용하다`로 정리해주신 부분이 크게 공감되네요. 좋은 코멘트를 남겨주셔서 감사합니다!
